images/converter/default.go

```diff
@@ -60,41 +58,95 @@ type ConvertHooks struct {
 
 // IndexConvertFuncWithHook is the convert func used by Convert with hook functions support.
 func IndexConvertFuncWithHook(layerConvertFunc ConvertFunc, docker2oci bool, platformMC platforms.MatchComparer, hooks ConvertHooks) ConvertFunc {
-	c := &defaultConverter{
+	return newDefaultWithOpts(convertOpts{
 		layerConvertFunc: layerConvertFunc,
 		docker2oci:       docker2oci,
 		platformMC:       platformMC,
-		diffIDMap:        make(map[digest.Digest]digest.Digest),
 		hooks:            hooks,
+	}).Convert
+}
+
+func NewDefault(opts ...Opt) (*DefaultConverter, error) {
+	var copts convertOpts
+	for _, o := range opts {
+		if err := o(&copts); err != nil {
+			return nil, err
+		}
+	}
+
+	return newDefaultWithOpts(copts), nil
+}
+
+func newDefaultWithOpts(copts convertOpts) *DefaultConverter {
+	if copts.platformMC == nil {
+		copts.platformMC = platforms.All
+	}
+
+	return &DefaultConverter{
+		convertOpts: copts,
 	}
-	return c.convert
 }
 
-type defaultConverter struct {
-	layerConvertFunc ConvertFunc
-	docker2oci       bool
-	platformMC       platforms.MatchComparer
-	diffIDMap        map[digest.Digest]digest.Digest // key: old diffID, value: new diffID
-	diffIDMapMu      sync.RWMutex
-	hooks            ConvertHooks
+type DefaultConverter struct {
+	convertOpts
+}
+
+type diffIDMap struct {
+	diffIDs map[digest.Digest]digest.Digest // key: old diffID, value: new diffID
+	mutex   sync.RWMutex
+}
+
+func makeRefKey(desc ocispec.Descriptor) (string, error) {
+	key, err := json.Marshal(desc)
+	if err != nil {
+		return "", err
+	}
+	return string(key), nil
+}
+
+// singleflightDo makes sure that only one conversion execution is
+// in-flight for a given descriptor at a time.
+func (c *DefaultConverter) singleflightDo(ctx context.Context, cs content.Store, desc ocispec.Descriptor, convert ConvertFunc) (*ocispec.Descriptor, error) {
+	if c.sg == nil {
+		return convert(ctx, cs, desc)
+	}
+
+	key, err := makeRefKey(desc)
+	if err != nil {
+		return nil, err
+	}
+
+	newDesc, err, _ := c.sg.Do(key, func() (interface{}, error) {
+		return convert(ctx, cs, desc)
```

Comment:
It seems that the invocation that skips a conversion doesn't have the lease on the result blob? If not, it'll result in the race where the blob gets GC-ed before the invocation (that skipped conversion) read it.

Reply:
@ktock Thanks for pointing it out, try to fix it in the fourth commit.